### PR TITLE
Expose a REST endpoint to download schemas

### PR DESCRIPTION
**Pull request by tomekl007**
WIP

TODO:
- [x] revisit media type (per https://github.com/stargate/stargate/pull/667#discussion_r574805538)

### CODE REVIEW

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java [hunk 1]
```diff
@@ -59,7 +51,7 @@ public void get(
       @Context HttpServletRequest httpRequest,
       @Suspended AsyncResponse asyncResponse) {
 
-    GraphQL graphql = getGraphl(namespace, httpRequest, asyncResponse);
+    GraphQL graphql = getGraphql(namespace, httpRequest, asyncResponse, graphqlCache);
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java:54]: It is not clear to me how it should work (and how it works now).
1. Is the `NamespaceResource` serving the schema based on version? If yes, why it needs to accept the query, variables, and other query parameters? Shouldn't it just return the schema for a given version?
A similar pattern is for `AdminResource` get method. Shouldn't we just return the `SchemaSource` from the new endpoint?

File: testing/src/test/java/io/stargate/it/http/graphfirst/AdminResourcesTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  @BeforeAll
+  public static void crateTestKeyspace() {
+    KEYSPACE = UUID.randomUUID().toString();
+    session.execute(
+        "CREATE KEYSPACE IF NOT EXISTS ks_1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1 };");
+  }
+
+  @Test
+  public void createTable() throws ExecutionException, InterruptedException {
```
Comment (tomekl007) [inline testing/src/test/java/io/stargate/it/http/graphfirst/AdminResourcesTest.java:29]: I am trying to write an IT test for admin resources to understand how schema retrieval works. I am trying to base it on the exiting cql-first `GraphqlTest`. 
From what I see, this test is pre-creating all schema via cql query from the `betterbotz.cql`. 
Regarding the new `graph-first` API my impression is that we should allow end-user to create everything via `graph-ql` queries only. Is that true?
If yes, can the new test use the same types, i.e: ` GetTableQuery.Table` or do we need to define new ones?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/SchemaFirstCache.java [hunk 3]
```diff
@@ -74,12 +76,13 @@ public GraphQL getAdminGraphql() {
...
-    return getGraphql(persistence.decorateKeyspaceName(namespace, headers));
+  public GraphQL getGraphql(String namespace, Map<String, String> headers, Optional<UUID> version)
+      throws Exception {
+    return getGraphql(persistence.decorateKeyspaceName(namespace, headers), version);
   }
 
-  private GraphQL getGraphql(String namespace) throws Exception {
-    SchemaSource latestSource = schemaSourceDao.getLatest(namespace);
+  private GraphQL getGraphql(String namespace, Optional<UUID> version) throws Exception {
+    SchemaSource latestSource = schemaSourceDao.getByVersion(namespace, version);
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/SchemaFirstCache.java:85]: One solution for exposing the schema to the end-user would be to rework this method to return the `GraphqlHolder`. The caller would be able to access the `SchemaSrouce` but also `GraphQL`. 
However, in the `SchemaSource` java-doc I see that:
```
 * <p>Usually this schema is partial, it will be completed (with default/generated types, directive
 * definitions...) to produce the actual one used at runtime.
```
so I am not sure if this is proper information that should be returned to a client.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/SchemaFirstCache.java [hunk 4]
```diff
@@ -74,12 +76,13 @@ public GraphQL getAdminGraphql() {
...
-  public GraphQL getGraphql(String namespace, Map<String, String> headers) throws Exception {
-    return getGraphql(persistence.decorateKeyspaceName(namespace, headers));
+  public GraphQL getGraphql(String namespace, Map<String, String> headers, Optional<UUID> version)
+      throws Exception {
+    return getGraphql(persistence.decorateKeyspaceName(namespace, headers), version);
   }
 
-  private GraphQL getGraphql(String namespace) throws Exception {
-    SchemaSource latestSource = schemaSourceDao.getLatest(namespace);
+  private GraphQL getGraphql(String namespace, Optional<UUID> version) throws Exception {
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/SchemaFirstCache.java:84]: I am not sure if we should modify this method to get the schema for a specific version or create a new method.
The reason for this concern is that this method uses the `graphqlHolders` map, which keeps only the last holder per namespace.
If customers will often request schema with non-lastest version, this "cache" will have a lot of miss (because elements will be constantly replaced), wdyt?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java (hunk 1 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java:54]: There are two aspects of the schema:
* the "raw" GraphQL string that the user passed to `deploySchema`. This is what we store in the `graphql_schema` table.
* the "runtime" `GraphQL` object that we use to execute queries. We build it by processing the raw schema, adding any generated types, binding the fetchers, etc. (see `SchemaProcessor`).

In this ticket, the goal is to expose the raw schema (so that users can download a previous version of their schema, modify it locally and re-deploy it).

> Shouldn't we just return the `SchemaSource` from the new endpoint?

Correct. `NamespaceResource` handles user queries on an already deployed schema, so it deals with the "runtime" schema. It's not the right place to look, instead the new REST resource should be more similar to `SingleSchemaFetcher`: call `SchemaSourceDao.get()` for the provided namespace and version, and then serve the `contents` field. You'll have to register a DAO instance in the Jersey environment in order to inject it in the resource, see `DropwizardServer`.
`SchemaFirstCache` builds its own DAO instance, so we might want to refactor to reuse a single one instead.

I haven't done a great job with naming. Maybe the `schema*` admin queries should have some notion of "raw" or "source" in the name, I guess that might be confusing for users too. Also I don't have a good name for the "non-admin" GraphQL schema, i.e. what gets deployed to a namespace (your question shows that `NamespaceResource` was a poor choice). I thought of "runtime", but that's not the best description because admin queries execute at runtime too. If you have suggestions I'd like to hear them.

File: testing/src/test/java/io/stargate/it/http/graphfirst/AdminResourcesTest.java (hunk 2 above)
Comment (olim7t) [reply testing/src/test/java/io/stargate/it/http/graphfirst/AdminResourcesTest.java:29]: `GraphqlTest` uses the Apollo client, this is good to have an example with a "real-life" GraphQL library, but adding new tests requires a lot of boilerplate. See the discussion on #270.
For our schema-first tests I think it will be faster to use the approach in `GraphqlJWTAuthTest`.

> Regarding the new `graph-first` API my impression is that we should allow end-user to create everything via `graph-ql` queries only. Is that true?

Yes. I think most of the time we won't need to run any CQL queries, the test will start with a `deploySchema` operation, and proceed from there. Except of course to cover specific database states (e.g. migration of an existing schema, `graphql_schema` exists but with the wrong columns, etc).

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java (hunk 1 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java:54]: The `SchemaSourceDao` accepts `DataStore` as an argument. We could remove it from the constructor, and add pass it as an argument to every method that needs it. However, as I see the `SchemaSourceDao` is stateless (besides the usage of the `dataStore`) so I think we can create a new instance of it wherever it's needed. Also, there are some code paths that do not have `DataStore` instance (i.e when `SchemaSourceDao` is called from *Resource),
wdyt?
Regarding naming, maybe:
- `RawSchema` and `ProcessedSchema`?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 5]
```diff
@@ -38,6 +50,27 @@ public Response getCqlDirectives() {
...
+        schemaSourceDao.getByVersion(namespace, Optional.ofNullable(version).map(UUID::fromString));
+
+    return Response.ok(schemaSource.getContents())
+        .header("Content-Disposition", "inline; filename=" + createFileName(schemaSource))
+        .build();
+  }
+
+  private String createFileName(SchemaSource schemaSource) {
+    return String.format(
+        "\"%s-%s.graphql\"", schemaSource.getNamespace(), schemaSource.getVersion());
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:71]: I think that it may be UX friendly if we will include both namespace and version in the schema name. Wdyt?

Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:71]: Yes, that's what I had in mind 👍

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 6]
```diff
@@ -38,6 +50,27 @@ public Response getCqlDirectives() {
...
 
+  @GET
+  @Path("/namespace/{namespaceName}.graphql")
+  @Produces(MediaType.TEXT_PLAIN)
+  public Response getSchema(
+      @PathParam("namespaceName") String namespace,
+      @QueryParam("version") String version,
+      @Context HttpServletRequest httpRequest)
+      throws Exception {
+    SchemaSource schemaSource =
```
Comment (dougwettlaufer) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:61]: What's this going to return if the namespace doesn't exist?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 7]
```diff
@@ -38,6 +50,27 @@ public Response getCqlDirectives() {
         .build();
   }
 
+  @GET
+  @Path("/namespace/{namespaceName}.graphql")
```
Comment (dougwettlaufer) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:54]: I know I'm late to the party but do we think "namespace" used in graphql will be confused with "namespace" in the docsapi?

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java [hunk 8]
```diff
@@ -49,4 +67,85 @@ public void shouldGetGraphQLDirectivesFile() throws IOException {
...
+        .contains((String.format("%s-%s.graphql", keyspaceId.toString(), DEPLOYED_SCHEMA_VERSION)));
+
+    String responseBody = getSchemaResponse.body().string();
+    assertThat(responseBody).isNotEmpty();
+    assertThat(responseBody).contains("type User");
+    assertThat(responseBody).contains("type Query");
+  }
+
+  @Test
+  public void shouldGetFileWithSchemaUsingVersionArgument(@TestKeyspace CqlIdentifier keyspaceId)
```
Comment (dougwettlaufer) [inline testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java:96]: Would be good to add a test for namespace that doesn't exist. What happens if I provide a valid keyspace that hasn't been graphqlified like `system`?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 9]
```diff
@@ -38,6 +50,27 @@ public Response getCqlDirectives() {
...
+  @Produces(MediaType.TEXT_PLAIN)
+  public Response getSchema(
```
Comment (dougwettlaufer) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:56]: All of the endpoints in here should require the usual auth headers as well.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 7 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:54]: I picked it because it's essentially the same thing: a thin abstraction over a CQL keyspace. But I don't mind changing it if you think we should use something different.

Btw you're not late, there's still time to improve naming and in fact I planned to do so before we merge the main PR (starting with the URIs).

Comment (dougwettlaufer) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:54]: Cool, I think that makes sense. We probably just need to be cognizant of the double meaning when documenting or discussing. `namespace` is probably a more familiar term for those using graphql as well.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 9 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:56]: 👍 , this method should do the same check as `SchemaFetcher`.

`getCqlDirectives()` returns static contents that will be the same for every Stargate installation, so I think it  can skip auth.

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java (hunk 8 above)
Comment (olim7t) [reply testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java:96]: +1 to add a test, as well as for existing namespace but wrong version.

From a quick look at the code it looks like the DAO returns `null` and that gets encoded as the response. I think a better behavior would be to return a 404.

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java [hunk 10]
```diff
@@ -49,4 +67,85 @@ public void shouldGetGraphQLDirectivesFile() throws IOException {
     assertThat(responseBody).isNotEmpty();
     assertThat(responseBody).contains("directive @cql_input");
   }
+
+  @Test
+  public void shouldGetFileWithSchema(@TestKeyspace CqlIdentifier keyspaceId) throws IOException {
+    // when get schema file
+    String url = String.format("%s:8080/graphqlv2/files/namespace/%s.graphql", host, keyspaceId);
```
Comment (olim7t) [inline testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java:74]: Can you use `ResourcePaths` instead of hard-coding `/graphqlv2/files`?
Paths are likely to change in the near future.

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java [hunk 11]
```diff
@@ -17,15 +17,33 @@
...
 import javax.ws.rs.core.MediaType;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
 import okhttp3.Response;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
 
+@ExtendWith(CqlSessionExtension.class)
 public class FilesResourceTest extends GraphqlITBase {
```
Comment (olim7t) [inline testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java:35]: Nit: could you name the parent class `GraphqlTestBase` instead?
The existing integration tests all use "Test", I think it's better to stay consistent.

**Review by olim7t — CHANGES_REQUESTED** [review]
The integration tests are getting hard to follow. Suggestions:
* rename `GraphqlTest` to `GraphqlApolloTest` to emphasize that it's the only one using the Apollo library
* follow the production package names
* I think `GraphqlJWTAuthTest` can now also extend `GraphqlITBase`

That would leave us with something like:
```
graphql/
  GraphqlTestBase
  TupleHelper
  cqlfirst/
    GraphqlApolloTest
    GraphqlJWTAuthTest
  schemafirst/
    FilesResourceTest
```

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java (hunk 10 above)
Comment (tomekl007) [reply testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java:74]: the `testing` module does not have a dependency on the `graphql` module, therefore the `ResourcePaths` is not available.

**Comment by tomekl007** [comment]
> The integration tests are getting hard to follow. Suggestions:
> 
> * rename `GraphqlTest` to `GraphqlApolloTest` to emphasize that it's the only one using the Apollo library
> * follow the production package names
> * I think `GraphqlJWTAuthTest` can now also extend `GraphqlITBase`
> 
> That would leave us with something like:
> 
> ```
> graphql/
>   GraphqlTestBase
>   TupleHelper
>   cqlfirst/
>     GraphqlApolloTest
>     GraphqlJWTAuthTest
>   schemafirst/
>     FilesResourceTest
> ```

Done, regarding:
>  I think `GraphqlJWTAuthTest` can now also extend `GraphqlITBase`
I don't think we should do it. The `GraphqlTestBase` uses the v2 API, whereas the `GraphqlJWTAuthTest` relies on the v1 API. Also, the  `GraphqlJWTAuthTest` uses a different strategy for generating `authTokens` and authorization.
The only common code is the `getHttpClient`. We can use it from this class since it is static in the `GraphqlTestBase`. Also, it is hard for me to refactor the `GraphqlJWTAuthTest` because it is not running on my local, getting:
```
org.testcontainers.containers.ContainerLaunchException: Timed out waiting for URL to be accessible (http://localhost:55007/auth/realms/master should return HTTP 200)
```

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlResourceBase.java [hunk 12]
```diff
@@ -338,4 +341,41 @@ protected static void replyWithGraphqlError(
...
+      } else {
+        return graphql;
+      }
+    } catch (Exception e) {
+      LOG.error("Unexpected error while accessing namespace {}", namespace, e);
+      replyWithGraphqlError(
+          Response.Status.NOT_FOUND, "Unexpected error while accessing namespace", asyncResponse);
+      return null;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlResourceBase.java:380]: Do you want a new empty class only for the purpose of being explicit? I've created a new `SchemaFirstResourceBase`

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 9 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:55]: Are you sure that we can do the same check as `SchemaFetcher#authorize`?
I see a problem with it: we don't have the `AuthenticationSubject` in the `FilesResource`, but it's needed for the `authorize` check. Should we construct the `ImmutableAuthenticationSubject` somehow? Should the `FilesResource` create the `AuthenticationSubject` from the `X-Cassandra-Token` header?

2nd question: 
I don't see an `authZ` in any other `grapqhl-first` resources: (`NamespaceResource`, `AdminResource`).
Are we missing the `authZ` there as well?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java (hunk 1 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java:53]: > However, as I see the `SchemaSourceDao` is stateless (besides the usage of the `dataStore`) so I think we can create a new instance of it wherever it's needed.

Thew way you did it in `DropwizardServer` looks good to me.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlResourceBase.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlResourceBase.java:380]: No, it will contain the namespace-related stuff, which is specific to schema-first.

[the remark this answers is not in the document]
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlResourceBase.java:380]: Hrmm, in fact I see that `getGraphql()` is only used by `NamespaceResource` now, probably a leftover from your initial prototype. We don't need a parent class anymore, I'll revert the changes.

File: testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java (hunk 10 above)
Comment (olim7t) [reply testing/src/test/java/io/stargate/it/http/graphfirst/FilesResourceTest.java]: Good point. Maybe we can extract the paths in the parent test class then. They'll still be duplicated, but only in one place.

**Comment by olim7t** [comment]
> The `GraphqlTestBase` uses the v2 API, whereas the `GraphqlJWTAuthTest` relies on the v1 API

In that case `GraphqlApolloTest` should not extend `GraphqlTestBase` either, it also uses the v1 API.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 9 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:55]: We do authentication and authorization in the REST API, it should be the same here.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 13]
```diff
@@ -38,9 +70,73 @@ public Response getCqlDirectives() {
...
+        isAuthorized(token, namespace, httpRequest)
+            ? schemaSourceDao.getByVersion(
+                namespace, Optional.ofNullable(version).map(UUID::fromString))
+            : null;
+    if (schemaSource == null) {
+      return Response.status(
+              Response.Status.NOT_FOUND.getStatusCode(),
+              String.format(
+                  "The schema for namespace %s and version %s does not exists, "
+                      + "or you're not authorized to read it.",
```
Comment (dougwettlaufer) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:93]: Not authorized should be a 401

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        new Request.Builder()
+            .get()
+            .addHeader("content-type", MediaType.TEXT_PLAIN)
+            .url(url)
+            .build();
+
+    Response getSchemaResponse = getHttpClient("wrong auth token").newCall(getRequest).execute();
+
+    // then
+    assertThat(getSchemaResponse.code()).isEqualTo(404);
```
Comment (dougwettlaufer) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java:159]: Should be a 401

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    String responseBody = getSchemaResponse.body().string();
+    assertThat(responseBody).isNotEmpty();
+    assertThat(responseBody).contains("type User");
+    assertThat(responseBody).contains("type Query");
+  }
+
+  @Test
+  public void shouldReturnNotFoundIfSchemaDoesNotExists() throws IOException {
+    // when get schema file
+    String url = String.format("%s:8080/graphqlv2/files/namespace/%s.graphql", host, "unknown");
```
Comment (dougwettlaufer) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java:127]: What about correct namespace and wrong version? Is that a 404 or a 400?

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java [hunk 16]
```diff
@@ -38,9 +70,73 @@ public Response getCqlDirectives() {
         .build();
   }
 
+  @GET
+  @Path("/namespace/{namespaceName}.graphql")
+  @Produces(MediaType.TEXT_PLAIN)
+  public Response getSchema(
+      @HeaderParam("X-Cassandra-Token") String token,
+      @PathParam("namespaceName") String namespace,
+      @QueryParam("version") String version,
```
Comment (dougwettlaufer) [inline graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:79]: Are versions user defined or an incremental number? Just wondering if this should be an int or if there should be some sort of other validation.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 9 above)
Comment (dougwettlaufer) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:76]: Is there an issue or TODO somewhere to also make sure this is added to `getCqlDirectives`?

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+            .build();
+
+    // when
+    Response response = httpClient.newCall(getRequest).execute();
+
+    // then
+    assertThat(response.code()).isEqualTo(200);
+    assertThat(response.body()).isNotNull();
+    String responseBody = response.body().string();
+    assertThat(responseBody).isNotEmpty();
```
Comment (dougwettlaufer) [inline testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java:68]: RestUtils does this already
https://github.com/stargate/stargate/blob/master/testing/src/test/java/io/stargate/it/http/RestUtils.java#L45-L64

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 16 above)
Comment (dougwettlaufer) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:79]: Oh, looks like a UUID? In that case should probably validate it and return an appropriate error message rather than bubbling up the 500.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 13 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:93]: Do we want to make that distinction though? If we return a 401, we're leaking the fact that the namespace and version exist.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 16 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:79]: Versions are timeuuids.

File: testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java (hunk 15 above)
Comment (olim7t) [reply testing/src/test/java/io/stargate/it/http/graphql/graphfirst/FilesResourceTest.java:127]: It will be a 404 as well. I'll add a test.

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 16 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:79]: > In that case should probably validate it and return an appropriate error message rather than bubbling up the 500

Indeed. I returned a 404 as well because versions are opaque strings as far as users are concerned.

**Comment by olim7t** [comment]
@tomekl007 I refactored the integration tests, I hope you don't mind, I want to merge this PR because I need the changes in #679.
I ended up ditching parent classes, in favor of utility classes like `RestUtils` (but with some state).

File: graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java (hunk 9 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/FilesResource.java:76]: `getCqlDirectives()` serves generic content, from a file that is publicly available in this github repo.

**Review by olim7t — APPROVED** [review]
